### PR TITLE
Introduce Validator plugin type

### PR DESCRIPTION
**Pull request by chuneuny-emily**
### Summary
Modularize Validator component to component and plugin

### How to test

python -m unittest -v tests/test_validator.py

### Checklist

- [x] I submit my changes into the `develop` branch
- [ ] I have added description of my changes into [CHANGELOG](https://github.com/openvinotoolkit/datumaro/blob/develop/CHANGELOG.md)
- [x] I have updated the [documentation](
  https://github.com/openvinotoolkit/datumaro/tree/develop/docs) accordingly
- [x] I have added tests to cover my changes
- [ ] I have [linked related issues](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/openvinotoolkit/datumaro/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2021 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

**Comment by chuneuny-emily** [comment]
Hi Maxim,
Can you review this PR so that it could be merged quickly? Please feel free to give comments if there are any feedbacks on implementation.

File: datumaro/components/validator.py [hunk 1]
```diff
@@ -18,21 +16,15 @@
...
-class Severity(Enum):
-    warning = auto()
-    error = auto()
+Severity = Enum('Severity', ['warning', 'error'])
 
-class TaskType(Enum):
-    classification = auto()
-    detection = auto()
-    segmentation = auto()
+TaskType = Enum('TaskType', ['classification', 'detection', 'segmentation'])
```
Comment (zhiltsov-max) [inline datumaro/components/validator.py:24]: Revert these lines.

File: datumaro/plugins/dataset_validator.py [hunk 2]
```diff
@@ -0,0 +1,733 @@
...
+from datumaro.components.cli_plugin import CliPlugin
+from datumaro.components.dataset import IDataset
+from datumaro.components.errors import MultiLabelAnnotations, NegativeLength
+from datumaro.components.extractor import AnnotationType, LabelCategories
+from datumaro.util import parse_str_enum_value
+
+from datumaro.cli.util import MultilineFormatter
+from datumaro.cli.util.project import generate_next_file_name, load_project
+
+class DatasetValidator(Validator, CliPlugin):
```
Comment (zhiltsov-max) [inline datumaro/plugins/dataset_validator.py:21]: I still can't understand the purpose of this class.

If it provides some shared bits of implementation - make it a base class for others. If it provides a common interface - put that code into the base class. Now it violates the interface, which it is supposed to implement, and can't be used the same way as other Validator classes. 

I see it the same way as extractors defined: 
IExtractor - abstract interface, no implementations (not necessary)
Extractor - some common and default definitions of methods
SourceExtractor - a base class for an often scenario
format extractors - which are format specific and inherit either Extractor or SourceExtractor

Here is the same:
Validator - a common interface or a base class
_TaskValidator - a base class for this kind of validators
Classification etc. Validator - specific classes.

They are used this way:
```
validator_type = env.validators[task]
extra_args = validator_type.parse_cmdline(...)
validator = validator_type(extra_args)
report = validator.validate(dataset)
```

File: datumaro/plugins/dataset_validator.py [hunk 3]
```diff
@@ -0,0 +1,733 @@
...
+
+    def __init__(self, task_type):
+        self.task_type = parse_str_enum_value(task_type, TaskType)
+
+    def set_extra_args(self, extra_args):
+        self.few_samples_thr = extra_args.get('few_samples_thr')
+        self.imbalance_ratio_thr = extra_args.get('imbalance_ratio_thr')
+        self.far_from_mean_thr = extra_args.get('far_from_mean_thr')
+        self.dominance_ratio_thr = extra_args.get('dominance_ratio_thr')
+        self.topk_bins = extra_args.get('topk_bins')
```
[the remark this answers is not in the document]
Comment (chuneuny-emily) [reply datumaro/plugins/dataset_validator.py]: I applied your feedback.

File: datumaro/components/validator.py (hunk 1 above)
Comment (chuneuny-emily) [reply datumaro/components/validator.py]: I applied your feedback.

File: datumaro/plugins/dataset_validator.py (hunk 2 above)
Comment (chuneuny-emily) [reply datumaro/plugins/dataset_validator.py]: I applied your feedback.

Comment (chuneuny-emily) [reply datumaro/plugins/dataset_validator.py]: Hi Maxim,
Thanks for your feedbacks. I applied all again. Please review it again.

File: tests/test_validator.py [hunk 4]
```diff
@@ -393,10 +392,10 @@ def test_check_multi_label_annotations(self):
         self.assertIsInstance(actual_reports[0], MultiLabelAnnotations)
 
 
-class TestDetectionValidator(TestValidatorTemplate):
+class TestDetection(TestValidatorTemplate):
     @classmethod
     def setUpClass(cls):
-        cls.validator = DetectionValidator(few_samples_thr=1,
+        cls.validator = Detection(few_samples_thr=1,
```
Comment (zhiltsov-max) [inline tests/test_validator.py:398]:
```suggestion
        cls.validator = DetectionValidator(few_samples_thr=1,
```

File: tests/test_validator.py [hunk 5]
```diff
@@ -363,10 +362,10 @@ def test_check_imbalanced_attribute(self):
             self.assertTrue(len(actual_reports) == 0)
 
 
-class TestClassificationValidator(TestValidatorTemplate):
+class TestClassification(TestValidatorTemplate):
     @classmethod
     def setUpClass(cls):
-        cls.validator = ClassificationValidator(few_samples_thr=1,
+        cls.validator = Classification(few_samples_thr=1,
```
Comment (zhiltsov-max) [inline tests/test_validator.py:368]:
```suggestion
        cls.validator = ClassificationValidator(few_samples_thr=1,
```

**Review by zhiltsov-max — COMMENTED** [review]
Looks good. The last thing I'd do is move the current Validator class into the plugin and give it `_TaskValidator` name or similar. Then, add a new `Validator` in the `components.validator` with just `validate`, `compute statistics` and `generate reports` definitions.

File: datumaro/cli/contexts/project/__init__.py [hunk 6]
```diff
@@ -816,19 +815,17 @@ def build_validate_parser(parser_ctor=argparse.ArgumentParser):
...
+    if args.subset_name is not None:
+        dataset = dataset.get_subset(args.subset_name)
+        dst_file_name += f'-{args.subset_name}'
 
-    extra_args = {}
-    from datumaro.components.validator import _Validator
-    extra_args = _Validator.parse_cmdline(args.extra_args)
-    validation_results = validate_annotations(dataset, task_type, **extra_args)
+    validator_type = project.env.validators[args.task_type]
+    extra_args = validator_type.parse_cmdline(args.extra_args)
```
[the remark this answers is not in the document]
Comment (chuneuny-emily) [reply datumaro/cli/contexts/project/__init__.py]: I updated it.

File: datumaro/components/validator.py [hunk 7]
```diff
@@ -1,13 +1,9 @@
 # Copyright (C) 2021 Intel Corporation
 #
 # SPDX-License-Identifier: MIT
-
+import numpy as np
 from copy import deepcopy
 from enum import Enum, auto
```
[the remark this answers is not in the document]
Comment (chuneuny-emily) [reply datumaro/components/validator.py]: I updated it.

File: datumaro/components/validator.py [hunk 8]
```diff
@@ -18,21 +14,23 @@
...
     FarFromAttrMean, OnlyOneAttributeValue)
 from datumaro.components.extractor import AnnotationType, LabelCategories
-from datumaro.components.cli_plugin import CliPlugin
 from datumaro.util import parse_str_enum_value
 
-
-class Severity(Enum):
-    warning = auto()
-    error = auto()
+Severity = Enum('Severity', ['warning', 'error'])
```
[the remark this answers is not in the document]
Comment (chuneuny-emily) [reply datumaro/components/validator.py]: I updated it.

File: datumaro/components/validator.py [hunk 9]
```diff
@@ -18,21 +14,23 @@
...
 class TaskType(Enum):
     classification = auto()
     detection = auto()
     segmentation = auto()
 
+class IValidator:
+    """
+    Basic interface for Validator
+    """
+    def validate(self, dataset: IDataset):
```
[the remark this answers is not in the document]
Comment (chuneuny-emily) [reply datumaro/components/validator.py]: I updated it.

File: datumaro/plugins/validators.py [hunk 10]
```diff
@@ -0,0 +1,702 @@
+# Copyright (C) 2021 Intel Corporation
+#
+# SPDX-License-Identifier: MIT
+
+import argparse
+import json
+import logging as log
+import numpy as np
+from copy import deepcopy
```
[the remark this answers is not in the document]
Comment (chuneuny-emily) [reply datumaro/plugins/validators.py]: I updated it.

File: tests/test_validator.py [hunk 11]
```diff
@@ -777,9 +776,9 @@ def test_validate_annotations_classification(self):
             self.assertEqual(actual_summary, expected_summary)
 
     @mark_requirement(Requirements.DATUM_GENERAL_REQ)
-    def test_validate_annotations_detection(self):
-        actual_results = validate_annotations(self.dataset, 'detection',
-            **self.extra_args)
+    def test_validate_detection(self):
+        validator = Detection(**self.extra_args)
```
[the remark this answers is not in the document]
Comment (chuneuny-emily) [reply tests/test_validator.py]: I updated it.

File: tests/test_validator.py [hunk 12]
```diff
@@ -720,9 +719,9 @@ class TestValidateAnnotations(TestValidatorTemplate):
             'topk_bins': 0.1,
         }
     @mark_requirement(Requirements.DATUM_GENERAL_REQ)
-    def test_validate_annotations_classification(self):
-        actual_results = validate_annotations(self.dataset, 'classification',
-            **self.extra_args)
+    def test_validate_classification(self):
+        validator = Classification(**self.extra_args)
```
[the remark this answers is not in the document]
Comment (chuneuny-emily) [reply tests/test_validator.py]: I updated it.

File: tests/test_validator.py [hunk 13]
```diff
@@ -560,10 +559,10 @@ def test_check_far_from_attr_mean(self):
         self.assertIsInstance(actual_reports[0], FarFromAttrMean)
 
 
-class TestSegmentationValidator(TestValidatorTemplate):
+class TestSegmentation(TestValidatorTemplate):
     @classmethod
     def setUpClass(cls):
-        cls.validator = SegmentationValidator(few_samples_thr=1,
+        cls.validator = Segmentation(few_samples_thr=1,
```
[the remark this answers is not in the document]
Comment (chuneuny-emily) [reply tests/test_validator.py]: I updated it.

File: tests/test_validator.py (hunk 4 above)
Comment (chuneuny-emily) [reply tests/test_validator.py]: I updated it.

File: tests/test_validator.py (hunk 5 above)
Comment (chuneuny-emily) [reply tests/test_validator.py]: I updated it.

**Comment by chuneuny-emily** [comment]
Hi Maxim,
I updated all feedbacks. Appreciate if you review it again. Thanks

**Review by zhiltsov-max — APPROVED** [review]
Thank you for the contribution!

**Pull request merged**